modules/rostests/kmtests/rtl/RtlStrSafe.c

```diff
@@ -0,0 +1,185 @@
+ *                  Copyright 2019 Colin Finck <colin@reactos.org>
+ */
+
+#define KMT_EMULATE_KERNEL
+#include <kmt_test.h>
+#include <ntstrsafe.h>
+#include <ntdef.h>
+#include <ndk/rtlfuncs.h>
+
+#define TESTAPI static void
+
+static const WCHAR FormatStringInts[] = L"%d %d %d";
+static const WCHAR FormatStringIntsResult[] = L"1 2 3";
+static const WCHAR FormatStringStrs[] = L"%s %s %s";
+
+
+TESTAPI
+Test_RtlUnicodeStringPrintf()
+{
+    NTSTATUS Status;
+    PWSTR pBuffer = NULL;
+    size_t BufferSize;
+    size_t EqualBytes;
+    UNICODE_STRING UsString;
+
+    KmtStartSeh();
+
+    /* STATUS_SUCCESS test */
+    BufferSize = 6 * sizeof(WCHAR);
+    pBuffer = KmtAllocateGuarded(BufferSize);
+    if (!pBuffer)
+        goto Cleanup;
+
+    RtlFillMemory(pBuffer, BufferSize, 0xAA);
+    RtlInitEmptyUnicodeString(&UsString, pBuffer, BufferSize);
+
+    Status = RtlUnicodeStringPrintf(&UsString, FormatStringInts, 1, 2, 3);
+    EqualBytes = RtlCompareMemory(UsString.Buffer, FormatStringIntsResult, sizeof(FormatStringIntsResult));
+    ok_eq_hex(Status, STATUS_SUCCESS);
+    ok_eq_size(EqualBytes, sizeof(FormatStringIntsResult));
+    ok_eq_uint(UsString.Length, sizeof(FormatStringIntsResult) - sizeof(WCHAR));
+    ok_eq_uint(UsString.MaximumLength, BufferSize);    
+
+    KmtFreeGuarded(pBuffer);
+    pBuffer = NULL;
+
+    /* STATUS_BUFFER_OVERFLOW tests */
+    BufferSize = 2 * sizeof(WCHAR);
+    pBuffer = KmtAllocateGuarded(BufferSize);
+    if (!pBuffer)
+        goto Cleanup;
+
+    RtlInitEmptyUnicodeString(&UsString, pBuffer, BufferSize);
+
+    Status = RtlUnicodeStringPrintf(&UsString, FormatStringStrs, L"AAA", L"BBB", L"CCC");
+    EqualBytes = RtlCompareMemory(UsString.Buffer, L"AA", BufferSize);
+    ok_eq_hex(Status, STATUS_BUFFER_OVERFLOW);
+    ok_eq_size(EqualBytes, BufferSize);
+    ok_eq_uint(UsString.Length, UsString.MaximumLength);
+
+    KmtFreeGuarded(pBuffer);
+    pBuffer = NULL;
+
+
+    BufferSize = 7 * sizeof(WCHAR);
+    pBuffer = KmtAllocateGuarded(BufferSize);
+    if (!pBuffer)
+        goto Cleanup;
+
+    RtlInitEmptyUnicodeString(&UsString, pBuffer, BufferSize);
+
+    Status = RtlUnicodeStringPrintf(&UsString, FormatStringStrs, L"0123", L"4567", L"89AB");
+    EqualBytes = RtlCompareMemory(UsString.Buffer, L"0123 45", BufferSize);
+    ok_eq_hex(Status, STATUS_BUFFER_OVERFLOW);
+    ok_eq_size(EqualBytes, BufferSize);
+    ok_eq_uint(UsString.Length, UsString.MaximumLength);
+
+    KmtFreeGuarded(pBuffer);
+    pBuffer = NULL;
+
+    // Note: RtlUnicodeStringPrintf returns STATUS_BUFFER_OVERFLOW here while RtlUnicodeStringPrintfEx returns STATUS_INVALID_PARAMETER!
+    // Documented on MSDN and verified with the Win10 version of ntstrsafe.h
+    RtlInitEmptyUnicodeString(&UsString, NULL, 0);
+    Status = RtlUnicodeStringPrintf(&UsString, FormatStringStrs, L"AAA", L"BBB", L"CCC");
+    ok_eq_hex(Status, STATUS_BUFFER_OVERFLOW);
+
+
+Cleanup:
+    if (pBuffer)
```

Comment:
The way this is used, there's no path inwhich `pBuffer` will end up being non-null here.
Though I understand if you want to leave this as "future-proof"

Reply:
Exactly. As C has no smart pointers that free resources for me under all circumstances, I always make `goto Cleanup` the only return path and free allocated resources there.